File: 697.Degree-of-an-Array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findShortestSubArray = function (nums) {
  const map = new Map();

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];

    if (!map.has(num)) {
      map.set(num, {
        count: 1,
        first: i,
        last: i,
      });
    } else {
      const obj = map.get(num);

      map.set(num, {
        ...obj,
        count: obj.count + 1,
        last: i,
      });
    }
  }

  let degree = -Infinity;
  let result = Infinity;

  for (const [num, obj] of map) {
    const { count, first, last } = obj;
    const length = last - first + 1;

    if (count > degree) {
      degree = count;
      result = length;
    } else if (count === degree) {
      result = Math.min(result, length);
    }
  }

  return result;
};

File: python/697.Degree-of-an-Array.py
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        count = {}
        first = {}

        result = len(nums)
        degree = 0

        for i, num in enumerate(nums):
            if num not in first:
                first[num] = i

            count[num] = count.get(num, 0) + 1

            if count[num] > degree:
                degree = count[num]
                result = i - first[num] + 1
            elif count[num] == degree:
                result = min(result, i - first[num] + 1)

        return result